doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -0,0 +1,149 @@
+
+## Non-goals
+
+- Identify all opportunities to use the cost model in the optimizer.
+- Refactor the optimizer to be focused on the cost model.
+- Implement dynamic re-optimization based on changes in input statistics.
+- Collect all relevant statistics in storage and make them available to the cost model.
+
+# Explanation
+[explanation]: #explanation
+
+## Proposal
+
+A cost model is an abstract interpretation of the query: rather than running the query on relations and getting new relation out, we run the query on a symbolic description of its input relations and get a symbolic description of the result out.
+
+The general approach can be the same as `mz_transform::Typecheck`: recursively walk the `MirRelationExpr` with some contextual information, returning a symbolic term characterizing the query.
+
+### Context information
+
+Traditional databases track cardinality data and statistics about various columns. We will be interested in other statistics, too. Some possible examples:
+
+- How frequently does data arrive?
+- Is the given input physically monotonic or append-only? (I.e., there are no retractions.)
+- At what rate do we expect new data to arrive?
+- Some kind of locality property?
+
+### Symbolic costs
+
+A symbolic cost is a polynomial over the features of the input relations. For example, a batch query that loops over relation `R` twice and then computes its cross product with relation `Q` will have cost `2|R| + |R|*|Q|`. A traditional database typically computes cost just in terms of overall batch query time, but our costs are multi-dimensional. In rough feasibility order:
+
+| Dimension          | Input features                                               | Outputs                         |
+| ------------------ | ------------------------------------------------------------ | ------------------------------- |
+| Rehydration time   | Source cardinality                                           | Time estimate                   |
+| Latency per update | Source cardinality, expected per-update delta on each source | Time estimate (per source)      |
+| Storage            | Source cardinality, existing indices, cardinality of groups  | Arrangements and size estimates |
+| Scalability        | Distribution of keys (e.g., cardinality of distinct keys)    | Data parallelism estimate       |
+
+The *rehydration time* dimension is effectively a classical batch database cost model: given cardinalities on the input relations, predict how much work a given plan has to do.
+
+The *latency per update* dimension is unique to our streaming setting. Different query plans will lead to different costs when different sources produce new data. If a calculus metaphor is helpful, you can think of the cost model here producing a partial derivative with respect to each input; alternatively, the cost model’s aggregate *latency per update* cost is the [Jacobian](https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant) of that query.
```

Comment:
I think another way to think about this is to consider *latency per update* to be the derivative of *rehydration time*. Maybe if we approach things like that we can derive the cost formula for *latency per update* by applying auto-grad on the cost formula for *rehydration time*. I think that for some operators we might have to manually hard code the derivative, but for others those would be as expected. For example, for `Filter` we will have:
```
R * s // rehydration time
```
and 
```
s // latency per update
```
where `R` is the number of input elements (including possible retractions) and `s` is the selectivity of the `Filter` predicates.

Reply:
This way of thinking makes intuitive sense to me, but I don't understand the LIR/low-level stuff well enough to know that it's true. It also doesn't seem to account for the notion of partial derivative: i.e., different sources may update at different rates.

Reply:
I'm not really sure if the derivative of the re-hydration time is the latency per update. For example, the current naive, non-incremental implementations of window functions recompute results for an entire window partition on every change. This means that if we get, say, 1000 elements in 1000 updates (i.e., 1-by-1), then it's about 1000*1000, but rehydration is different: we get just one update with 1000 elements, and there are not recomputations.